### DISCUSSION TITLE
Download images

### DISCUSSION THREAD

**Discussion by arnaldop** [coddingtonbear/obsidian-web#92] [Ideas]
It would be great if we could also download the images within a page.
We should be able to configure the root directory for images.
Downloaded content should link to the downloaded files rather than their live versions.

Thanks for the great plugin!!!

**Comment by JroachAu** [comment]
Yes! Came here to look / post for this request. Primarily important when saving a page where it's behind a user account and the image URL isn't accessible. So instead of grabbing the URL of the image, it takes the image instead.

**Reply by Tback1** [reply]
I'm also in favor of including this feature. Recently, a website where I excerpted an article for a long time put the image into the CDN under its subdomain, and the worst thing is that to access the image above, first, you need to construct a link to convert the png into webp through the frame, and second, only the web page with the main domain name of the website can access the image resources under the CDN, otherwise it will be 403 processing. As a result, any part of the excerpt now needs to be picked up by yourself, copied and pasted the image, or downloaded in bulk and manually inserted.

**Comment by coddingtonbear** [comment]
Previously, I'd said "No" to this request (see https://github.com/coddingtonbear/obsidian-web/discussions/7), but I hadn't considered that you might be viewing a page requiring authentication and might want to take advantage of your browser's session to fetch whatever images with its credentials.

So, I guess I'm a bit less opposed to somebody adding a feature like this if they wanted, but if somebody does want to contribute such a feature, they should probably start a discussion here so we can come to an understanding of what those changes could/might look like and what I'd require before accepting such a Pull Request.  I'm afraid this particular feature is not something I would want to use in my own environment personally, so I'm not motivated enough to take on that work myself, but I'd be up for accepting a contribution if it was high-quality enough.

**Reply by JroachAu** [reply]
I'm a script/py/bash/ps writer, not what I would consider a proper developer. Any suggestions on where to start, particularly because I'd be working on the Chrome Extension and an obsidian extension. Irrespective, I'm pretty keen to help as it will save me time right now.

**Reply by coddingtonbear** [reply]
You might be selling yourself short, and you're welcome to give it a shot, but I'm afraid this sort of feature will require rather a lot of knowledge across a handful of domains.  Not only would you need to figure out a sane way of building a UI around configuring such a feature, but you'd need to figure out a way of gathering the page's images from within the extension's content script, sending them to the background script so they can be sent to Obsidian, sending them from that background script to obsidian and placing them in some sane spot in your vault, and then re-writing whatever page snapshot to reference the new image locations.  I wouldn't say it's *hard* exactly, but there are a lot of moving parts that would have to be carefully used or crafted.

**Comment by JroachAu** [comment]
From my perspective:
Process:
- USER selects content OR USER grabs whole page (i.e. 'Capture page snapshot')
- Images within are copied over as 'copy image' (current behaviour: 'copy image address')
- When the *.md file is created, the images would need to be copied as their image name and linked to within the md.

Options:
- [tickbox] Copy images as the Image file (default unticked)
- Suboptions:
-> [folder path] Image save location (default blank). if blank, same directory as the markdown file created, '\': home directory, or otherwise entered.

----
Other considerations? @arnaldop

**Reply by Tback1** [reply]
I guess most of the settings can refer to "Local Image Plus", just change the flow from the source to GET change to copy directly from the browser buffer

**Comment by sinkr** [comment]
One day I will be able to dump Evernote when these Markdown clippers catch up and sufficiently render a webpage in MD based on how it looks in browser.

I know the end result is possible, because I can take my Evernote notes, export them to `.enex` files and import them into Obsidian using the community 'Importer' plugin and get the images and a sufficiently decent render of the original.

It's just sad that there has to be all this in the middle.

**Comment by Tback1** [comment]
Roughly another solution is to add a reference to the obsidian access, after all, there is a relevant source URL in yaml when crawling, as long as the reference is the source address when accessing the network image, it can be accessed normally, and then downloaded to the local through the LocalImage plugin

**Reply by coddingtonbear** [reply]
That was my first instinct, too, @Tback1, but I'm afraid you're not considering images that require authentication. This happens transparently in your browser, but when your browser requests an image being shown on a particular page, it will include in the request the relevant cookies stored for the domain hosting the image, and the server serving that image can use those cookies for confirming that you are authorized to retrieve the image.

It's probably the case that *most* of the time images don't require authentication, but unfortunately there are a lot of situations in which authentication might be required, and in such cases it probably does make sense to grab the image from within the browser itself as a result.
